src/ADAL.PCL/Telemetry.cs

```diff
@@ -0,0 +1,112 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory
+{
+    public class Telemetry
+    {
+        private readonly static Telemetry Instance = new Telemetry();
+
+        public static Telemetry GetInstance()
+        {
+            return Instance;
+        }
+
+        DefaultDispatcher Dispatcher = null ;
+
+        private IDictionary<Tuple<string, string>,string> EventTracking = new ConcurrentDictionary<Tuple<string, string>,string>();
+
+        internal string RegisterNewRequest()
+        {
+            return new Guid().ToString();
+        }
+
+        public void RegisterDispatcher(IDispatcher dispatcher, bool aggregationRequired)
+        {
+            if (aggregationRequired)
+            {
+                Dispatcher = new DefaultDispatcher(dispatcher);
+            }
+            else
+            {
+                Dispatcher = new AggregatedDispatcher(dispatcher);
+            }
+        }
+
+        internal void StartEvent(string requestId, string eventName)
+        {
+            EventTracking.Add(new Tuple<string, string>(requestId,eventName),DateTime.UtcNow.ToString());
+        }
+
+        internal void StopEvent(string requestId, EventsBase Event,string eventName)
+        {
+            string value;
+            List<Tuple<string, string>> listEvent = Event.GetEvents(eventName);
+            if (EventTracking.
+                TryGetValue(new Tuple<string, string>(requestId, eventName), out value))
+            {
+                DateTime startTime = DateTimeOffset.Parse(value).UtcDateTime;
+                System.TimeSpan diff1 = DateTime.UtcNow.Subtract(startTime);
+                //Add the response time to the list
+                listEvent.Add(new Tuple<string, string>("response_time",diff1.ToString()));
+                //Adding event name to the start of the list
+                listEvent.Insert(0, new Tuple<string, string>("EventName",eventName));
+                //Remove the event from the tracking Map
+                EventTracking.Remove(new Tuple<string, string>(requestId, eventName));
+            }
+            Dispatcher.Receive(requestId,Event);
+            //remove it from the map
+            //time calculation
+            //add the response time
```

Comment:
I have already coded them above receive. The comments were mentioned before I started.removed them.